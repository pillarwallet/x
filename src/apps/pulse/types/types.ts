export type TransactionStatusState =
  | 'Starting Transaction'
  | 'Transaction Pending'
  | 'Transaction Complete'
  | 'Transaction Failed';

export type TransactionStep =
  | 'Submitted'
  | 'Pending'
  | 'ResourceLock'
  | 'Completed';

export type StepStatus = 'completed' | 'pending' | 'failed' | 'inactive';

export interface TokenInfo {
  symbol: string;
  name: string;
  logo: string;
  address?: string;
}

export interface SellOffer {
  tokenAmountToReceive: number;
  minimumReceive: number;
}

export interface PayingToken {
  totalUsd: number;
  name: string;
  symbol: string;
  logo: string;
  actualBal: string;
  totalRaw: string;
  chainId: number;
  address: string;
}

export interface TokenDetails {
  symbol: string;
  name: string;
  address: string;
  chainId: number;
  amount: string;
  logo: string;
  type: 'BUY_TOKEN' | 'SELL_TOKEN';
}

export interface BuyModeDetails {
  usdAmount: string;
  payingTokens: PayingToken[];
  totalPayingUsd: number;
}

export interface TransactionStatusConfig {
  icon: string;
  containerClasses: string;
  iconClasses: string;
  color: string;
}

export interface ButtonConfig {
  bgColor: string;
  textColor: string;
  borderColor: string;
  label: string;
}

export interface ProgressStepConfig {
  label: string;
  order: number;
}

export interface TransactionStatusProps {
  closeTransactionStatus: () => void;
  userOpHash: string;
  chainId: number;
  gasFee?: string;
  // Transaction data from PreviewSell/PreviewBuy
  isBuy?: boolean;
  sellToken?: TokenInfo | null;
  buyToken?: TokenInfo | null;
  tokenAmount?: string;
  sellOffer?: SellOffer | null;
  payingTokens?: PayingToken[];
  usdAmount?: string;
  // Externalized polling state
  currentStatus: TransactionStatusState;
  errorDetails: string;
  submittedAt?: Date;
  pendingCompletedAt?: Date;
  blockchainTxHash?: string;
  resourceLockTxHash?: string;
  completedTxHash?: string;
  completedChainId?: number;
  resourceLockChainId?: number;
  resourceLockCompletedAt?: Date;
  isResourceLockFailed?: boolean;
}

export interface TransactionDetailsProps {
  onDone: () => void;
  userOpHash: string;
  chainId: number;
  status: TransactionStatusState;
  // Transaction data from PreviewSell/PreviewBuy
  isBuy?: boolean;
  sellToken?: TokenInfo | null;
  buyToken?: TokenInfo | null;
  tokenAmount?: string;
  sellOffer?: SellOffer | null;
  payingTokens?: PayingToken[];
  usdAmount?: string;
  submittedAt?: Date;
  pendingCompletedAt?: Date;
  resourceLockCompletedAt?: Date;
  txHash?: string;
  gasFee?: string;
  errorDetails?: string;
  resourceLockTxHash?: string;
  completedTxHash?: string;
  resourceLockChainId?: number;
  completedChainId?: number;
  isResourceLockFailed?: boolean;
}

export interface TransactionInfoProps {
  status: TransactionStatusState;
  userOpHash: string;
  txHash?: string;
  chainId: number;
  gasFee?: string;
  completedAt?: Date;
  // Buy-specific fields
  isBuy?: boolean;
  resourceLockTxHash?: string;
  resourceLockChainId?: number;
  completedTxHash?: string;
  completedChainId?: number;
}

export interface ProgressStepProps {
  step: TransactionStep;
  status: StepStatus;
  label: string;
  isLast?: boolean;
  showLine?: boolean;
  lineStatus?: StepStatus;
  timestamp?: string | React.ReactNode;
}

export interface TransactionErrorBoxProps {
  technicalDetails?: string;
}

export interface UseClickOutsideOptions {
  ref: React.RefObject<HTMLElement>;
  callback: () => void;
  condition: boolean;
}

export interface UseKeyboardNavigationOptions {
  onEscape: () => void;
  onEnter?: () => void;
  enabled?: boolean;
}

export interface TechnicalDetails {
  transactionType: 'BUY' | 'SELL';
  transactionHash: string;
  hashType: 'bidHash' | 'userOpHash';
  chainId: number;
  status: TransactionStatusState;
  timestamp: string;
  accountAddress: string;
  token: TokenDetails | null;
  sellOffer: SellOffer | null;
  buyMode: BuyModeDetails | null;
  transactionHashes: {
    [key: string]: string;
  };
  chains: {
    mainChainId: number;
    resourceLockChainId: string | number;
    completedChainId: string | number;
  };
  timestamps: {
    [key: string]: string;
  };
  error: {
    details: string;
    isResourceLockFailed: boolean;
    failureStep: string;
  };
  gas: {
    fee: string;
  };
  stepStatus: {
    [key: string]: StepStatus;
  };
}
