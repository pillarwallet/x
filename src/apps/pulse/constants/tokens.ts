export const STABLE_CURRENCIES = [
  { chainId: 10, address: '0x0b2C639c533813f4Aa9D7837CAf62653d097Ff85' }, // USDC on Optimism
  { chainId: 137, address: '0x3c499c542cEF5E3811e1192ce70d8cC03d5c3359' }, // USDC on Polygon
  { chainId: 8453, address: '0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913' }, // USDC on Base
  { chainId: 42161, address: '0xaf88d065e77c8cC2239327C5EDb3A432268e5831' }, // USDC on Arbitrum
  { chainId: 56, address: '0x8AC76a51cc950d9822D68b83fE1Ad97B32Cd580d' }, // USDC on BNB Smart Chain
  { chainId: 100, address: '0x2a22f9c3b484c3629090FeED35F17Ff8F88f76F0' }, // USDC on Gnosis
];
