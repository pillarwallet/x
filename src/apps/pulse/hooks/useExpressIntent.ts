import { useEffect, useState } from 'react';
import {
  ExpressIntentResponse,
  UserIntent,
} from '@etherspot/intent-sdk/dist/cjs/sdk/types/user-intent-types';
import useIntentSdk from './useIntentSdk';

export default function useExpressIntent(userIntent: UserIntent | null) {
  const [isLoading, setIsLoading] = useState<boolean>(false);
  const [expressIntentResponse, setExpressIntentResponse] =
    useState<ExpressIntentResponse | null>(null);
  const { intentSdk } = useIntentSdk();
  const [refetch, setRefetch] = useState(false);

  useEffect(() => {
    if (intentSdk && userIntent) {
      setIsLoading(true);
      intentSdk
        ?.expressIntent(userIntent)
        .then((response) => {
          setExpressIntentResponse(response);
          setIsLoading(false);
        })
        .catch((error) => {
          console.error('Error in expressIntent: ', error);
          setIsLoading(false);
          setExpressIntentResponse(null);
        });
    } else {
      setExpressIntentResponse(null);
      setIsLoading(false);
    }
  }, [userIntent, intentSdk, refetch]);

  return {
    isLoading,
    expressIntentResponse,
    refetch,
    setRefetch,
  };
}
