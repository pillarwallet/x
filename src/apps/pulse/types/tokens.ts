type ParsedToken = {
  name: string;
  symbol: string;
  markrtCap: number;
  liquidity: number;
  volume: number;
  logo: string;
  price: number;
  priceChange: number;
  blockchain: string;
}

type SelectedToken = {
  name: string;
  symbol: string;
  usdValue: string;
  logo: string;
  dailyPriceChange: number;
  chainId: number;
  decimals: number;
  address: string;
}