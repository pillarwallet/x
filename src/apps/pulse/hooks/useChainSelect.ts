import { useState } from 'react';
import { MobulaChainNames } from '../utils/constants';

export default function useChainSelect() {
  const [chains, setChains] = useState<MobulaChainNames>(MobulaChainNames.All);

  return {
    chains,
    setChains,
  };
}
