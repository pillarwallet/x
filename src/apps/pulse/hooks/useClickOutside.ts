import { useEffect } from 'react';
import { UseClickOutsideOptions } from '../types/types';

/**
 * Custom hook for handling click outside events
 */
export const useClickOutside = ({
  ref,
  callback,
  condition,
}: UseClickOutsideOptions) => {
  useEffect(() => {
    const handleClickOutside = (event: MouseEvent) => {
      if (
        ref.current &&
        !ref.current.contains(event.target as Node) &&
        condition
      ) {
        callback();
      }
    };

    document.addEventListener('mousedown', handleClickOutside);

    return () => {
      document.removeEventListener('mousedown', handleClickOutside);
    };
  }, [ref, callback, condition]);
};
