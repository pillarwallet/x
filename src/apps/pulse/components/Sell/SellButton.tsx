import { Dispatch, SetStateAction } from 'react';
import { TailSpin } from 'react-loader-spinner';

// hooks
import { SellOffer } from '../../hooks/useRelaySell';

// types
import { SelectedToken } from '../../types/tokens';

// components
import HighDecimalsFormatted from '../../../pillarx-app/components/HighDecimalsFormatted/HighDecimalsFormatted';

// utils
import { limitDigitsNumber } from '../../../../utils/number';

function getButtonText(
  isLoading: boolean,
  isInitialized: boolean,
  selectedToken: SelectedToken | null,
  tokenAmount: string,
  sellOffer?: SellOffer | null,
  isDisabled?: boolean
) {
  if (!isInitialized) {
    return 'Initializing...';
  }

  if (isLoading) {
    return (
      <div className="flex items-center justify-center">
        <TailSpin color="#FFFFFF" height={15} width={15} />
      </div>
    );
  }

  const amount = parseFloat(tokenAmount);
  if (
    !isDisabled &&
    !Number.isNaN(amount) &&
    amount > 0 &&
    sellOffer?.tokenAmountToReceive
  ) {
    const limitedAmount = limitDigitsNumber(amount);
    const limitedUsdcAmount = limitDigitsNumber(sellOffer.tokenAmountToReceive);

    return (
      <div className="flex items-center justify-center gap-1">
        <span>Sell</span>
        <HighDecimalsFormatted
          value={limitedAmount}
          styleNumber="text-white text-sm"
          styleZeros="text-white/70 text-xs"
        />
        <span>{selectedToken?.symbol}</span>
        <span>for</span>
        <HighDecimalsFormatted
          value={limitedUsdcAmount}
          moneySymbol=""
          styleNumber="text-white text-sm"
          styleZeros="text-white/70 text-xs"
        />
        <span>USDC</span>
      </div>
    );
  }

  return selectedToken?.symbol ? `Sell ${selectedToken.symbol}` : 'Sell';
}

export interface SellButtonProps {
  token: SelectedToken | null;
  tokenAmount: string;
  notEnoughLiquidity: boolean;
  setPreviewSell: Dispatch<SetStateAction<boolean>>;
  setSellOffer: Dispatch<SetStateAction<SellOffer | null>>;
  sellOffer: SellOffer | null;
  isLoadingOffer: boolean;
  isInitialized: boolean;
}

const SellButton = (props: SellButtonProps) => {
  const {
    token,
    tokenAmount,
    notEnoughLiquidity,
    setPreviewSell,
    setSellOffer,
    sellOffer,
    isLoadingOffer,
    isInitialized,
  } = props;

  const handleSellClick = () => {
    if (sellOffer) {
      setSellOffer(sellOffer);
      setPreviewSell(true);
    }
  };

  const isDisabled = () => {
    return (
      isLoadingOffer ||
      !token ||
      !(parseFloat(tokenAmount) > 0) ||
      !sellOffer ||
      sellOffer.tokenAmountToReceive === 0 ||
      notEnoughLiquidity ||
      !isInitialized
    );
  };

  return (
    <button
      className="flex-1 items-center justify-center font-medium text-sm"
      onClick={handleSellClick}
      disabled={isDisabled()}
      style={{
        backgroundColor: isDisabled() ? '#29292F' : '#8A77FF',
        color: isDisabled() ? 'grey' : '#FFFFFF',
        borderRadius: 8,
      }}
      type="button"
      data-testid="pulse-sell-button"
    >
      <span>
        {getButtonText(
          isLoadingOffer,
          isInitialized,
          token,
          tokenAmount,
          sellOffer,
          isDisabled()
        )}
      </span>
    </button>
  );
};

export default SellButton;
