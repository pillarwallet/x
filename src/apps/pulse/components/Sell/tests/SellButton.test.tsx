/* eslint-disable react/jsx-props-no-spreading */
/* eslint-disable @typescript-eslint/no-explicit-any */
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { vi } from 'vitest';

// hooks
import { SellOffer } from '../../../hooks/useRelaySell';

// types
import { SelectedToken } from '../../../types/tokens';

// components
import SellButton from '../SellButton';

const mockToken: SelectedToken = {
  name: 'Test Token',
  symbol: 'TEST',
  logo: 'test-logo.png',
  usdValue: '100.00',
  dailyPriceChange: 0.05,
  chainId: 1,
  decimals: 18,
  address: '0x1234567890123456789012345678901234567890',
};

const mockSellOffer: SellOffer = {
  tokenAmountToReceive: 50.0,
  offer: {
    errors: undefined,
    fees: undefined,
    breakdown: undefined,
    details: undefined,
    error: undefined,
    refunded: undefined,
    steps: [],
    request: undefined,
  },
};

const mockProps = {
  setPreviewSell: vi.fn(),
  setSellOffer: vi.fn(),
  notEnoughLiquidity: false,
  isLoadingOffer: false,
  isInitialized: true,
};

describe('<SellButton />', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('renders correctly and matches snapshot', () => {
    const tree = renderer
      .create(
        <SellButton
          {...mockProps}
          token={mockToken}
          tokenAmount="10"
          sellOffer={mockSellOffer}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('renders correct text', () => {
    it('with token and offer available', () => {
      render(
        <SellButton
          {...mockProps}
          token={mockToken}
          tokenAmount="10"
          sellOffer={mockSellOffer}
        />
      );

      expect(screen.getByText('Sell')).toBeInTheDocument();
      expect(screen.getByText('10')).toBeInTheDocument();
      expect(screen.getByText('TEST')).toBeInTheDocument();
      expect(screen.getByText('for')).toBeInTheDocument();
      expect(screen.getByText('50')).toBeInTheDocument();
      expect(screen.getByText('USDC')).toBeInTheDocument();
    });

    it('with token only when no offer', () => {
      render(
        <SellButton
          {...mockProps}
          token={mockToken}
          tokenAmount="10"
          sellOffer={null}
        />
      );

      expect(screen.getByText('Sell TEST')).toBeInTheDocument();
    });

    it('shows "Sell" when no token selected', () => {
      render(
        <SellButton
          {...mockProps}
          token={null}
          tokenAmount="10"
          sellOffer={mockSellOffer}
        />
      );

      expect(screen.getByRole('button')).toHaveTextContent('Sell');
    });

    it('shows "Initializing..." when not initialized', () => {
      render(
        <SellButton
          {...mockProps}
          token={mockToken}
          tokenAmount="10"
          sellOffer={mockSellOffer}
          isInitialized={false}
        />
      );

      expect(screen.getByText('Initializing...')).toBeInTheDocument();
    });
  });

  describe('button states', () => {
    it('is enabled when all conditions are met', () => {
      render(
        <SellButton
          {...mockProps}
          token={mockToken}
          tokenAmount="10"
          sellOffer={mockSellOffer}
        />
      );

      const button = screen.getByRole('button');
      expect(button).not.toBeDisabled();
      expect(button).toHaveStyle({
        backgroundColor: '#8A77FF',
        color: '#FFFFFF',
      });
    });

    it('is disabled when conditions not met', () => {
      const { rerender } = render(
        <SellButton
          {...mockProps}
          token={null}
          tokenAmount="10"
          sellOffer={mockSellOffer}
        />
      );

      let button = screen.getByRole('button');
      expect(button).toBeDisabled();
      expect(button).toHaveStyle({
        backgroundColor: '#29292F',
        color: 'grey',
      });

      rerender(
        <SellButton
          {...mockProps}
          token={mockToken}
          tokenAmount="0"
          sellOffer={mockSellOffer}
        />
      );

      button = screen.getByRole('button');
      expect(button).toBeDisabled();

      rerender(
        <SellButton
          {...mockProps}
          token={mockToken}
          tokenAmount="10"
          sellOffer={null}
        />
      );

      button = screen.getByRole('button');
      expect(button).toBeDisabled();

      rerender(
        <SellButton
          {...mockProps}
          token={mockToken}
          tokenAmount="10"
          sellOffer={mockSellOffer}
          notEnoughLiquidity
        />
      );

      button = screen.getByRole('button');
      expect(button).toBeDisabled();
    });
  });

  it('calls handlers when clicked', () => {
    render(
      <SellButton
        {...mockProps}
        token={mockToken}
        tokenAmount="10"
        sellOffer={mockSellOffer}
      />
    );

    screen.getByRole('button').click();

    expect(mockProps.setPreviewSell).toHaveBeenCalledWith(true);
    expect(mockProps.setSellOffer).toHaveBeenCalledWith(mockSellOffer);
  });
});
