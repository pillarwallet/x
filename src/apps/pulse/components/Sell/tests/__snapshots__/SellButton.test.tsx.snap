// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<SellButton /> > renders correctly and matches snapshot 1`] = `
<button
  className="flex-1 items-center justify-center"
  disabled={false}
  onClick={[Function]}
  style={
    {
      "backgroundColor": "#8A77FF",
      "borderRadius": 8,
      "color": "#FFFFFF",
    }
  }
  type="button"
>
  Sell 10.0000 TEST for 50.0000 USDC
</button>
`;
